ValueError: Invalid data-type for array with Python 3.11

While installing pyerfa from source (as an Astropy dependency) on Python 3.11, I get this error with the ufunc calls. 
For exemple, running tests from the installed package results in these errors: 
```
❯ pytest ~/.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests

=================================== short test summary info ====================================
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::test_erfa_wrapper - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::test_errwarn_reporting - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::test_vector_inouts - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::test_rz - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::test_float32_input - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::TestLeapSeconds::test_set_leap_seconds - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::TestLeapSeconds::test_validation[table0-January] - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::TestLeapSeconds::test_validation[table1-jump] - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::TestLeapSeconds::test_validation[table2-dimension] - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_erfa.py::TestLeapSeconds::test_update_leap_seconds - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_ab - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_c2ibpn - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_c2ixy - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_c2tcio - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_c2teqx - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_cal2jd - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_dat - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_eform - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_eors - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_epj - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_gc2gd - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_gc2gde - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_gd2gc - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_gst06 - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_ld - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_ldsun - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_num00b - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_num06a - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_pmpx - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_ppsp - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_pr00 - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_rx - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_ry - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_rz - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_sxp - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_tporv - ValueError: Invalid data-type for array
FAILED ../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py::test_tpstv - ValueError: Invalid data-type for array
==================== 37 failed, 236 passed, 5 xfailed, 22 warnings in 3.10s ====================
```

Failure example:
```
__________________________________________ test_tporv __________________________________________

    def test_tporv():
        v = np.empty((3,), float)
        vz1 = np.empty((3,), float)
        vz2 = np.empty((3,), float)
        xi = -0.03
        eta = 0.07
        ra = 1.3
        dec = 1.5
        v = erfa_ufunc.s2c(ra, dec)
>       vz1, vz2, n = erfa_ufunc.tporv(xi, eta, v)
E       ValueError: Invalid data-type for array

../../../.pyenv/versions/3.11.0/lib/python3.11/site-packages/erfa/tests/test_ufunc.py:3511: ValueError
```

Astropy's CI on Python 3.11 is not failing, so I don't know, maybe something due to the GCC version ?

```
❯ gcc --version                                           
gcc (GCC) 12.2.0
```

When building from source I get many "incompatible pointer type" warnings, and a "dangling pointer" one:
```
      |         void (*)(char **, npy_intp *, npy_intp *, void *) {aka void (*)(char **, long int *, long int *, void *)}
  erfa/ufunc.c:14046:9: note: expected ‘PyUFuncGenericFunction’ {aka ‘void (*)(char **, const long int *, const long int *, void *)’} but argument is of type ‘void (*)(char **, npy_intp *, npy_intp *, void *)’ {aka ‘void (*)(char **, long int *, long int *, void *)’}
  erfa/ufunc.c:14080:9: warning: passing argument 3 of ‘(int (*)(PyUFuncObject *, PyArray_Descr *, void (*)(char **, const npy_intp *, const npy_intp *, void *), PyArray_Descr **, void *))*(PyUFunc_API + 328)’ from incompatible pointer type [-Wincompatible-pointer-types]
  14080 |         ufunc_loop_sxpv, dtypes, NULL);
        |         ^~~~~~~~~~~~~~~
        |         |
        |         void (*)(char **, npy_intp *, npy_intp *, void *) {aka void (*)(char **, long int *, long int *, void *)}
  erfa/ufunc.c:14080:9: note: expected ‘PyUFuncGenericFunction’ {aka ‘void (*)(char **, const long int *, const long int *, void *)’} but argument is of type ‘void (*)(char **, npy_intp *, npy_intp *, void *)’ {aka ‘void (*)(char **, long int *, long int *, void *)’}
  erfa/ufunc.c: In function ‘ErfaUFuncTypeResolver’:
  erfa/ufunc.c:9729:16: warning: dangling pointer ‘types’ to ‘types_array’ may be used [-Wdangling-pointer=]
   9729 |         return set_ufunc_loop_data_types(ufunc, operands, out_dtypes,
        |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   9730 |                                          types, dtypes);
        |                                          ~~~~~~~~~~~~~~
  erfa/ufunc.c:9715:13: note: ‘types_array’ declared here
   9715 |         int types_array[NPY_MAXARGS];
        |             ^~~~~~~~~~~
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,3 +1,8 @@
+2.0.0.4 (unreleased)
+====================
+
+- Fix dangling pointer leading to unexpected behavior with ``-O3``.
+
 2.0.0.3 (2023-03-22)
 ====================
 

--- a/erfa/ufunc.c.templ
+++ b/erfa/ufunc.c.templ
@@ -556,6 +556,7 @@ static int ErfaUFuncTypeResolver(PyUFuncObject *ufunc,
 {
     int *types;
     PyArray_Descr **dtypes;
+    int types_array[NPY_MAXARGS];
 
     if (ufunc->userloops) {
         Py_ssize_t unused_pos = 0;
@@ -577,7 +578,6 @@ static int ErfaUFuncTypeResolver(PyUFuncObject *ufunc,
     }
     else {
         npy_intp j;
-        int types_array[NPY_MAXARGS];
 
         if (ufunc->ntypes != 1) {
             goto fail;

